File: nested_sampling.py
"""
Simple code that implements the nested sampling
to compute the evidence D-dimensiona gaussian.
The posterior distributions of the parameters are also calculated.
It was always assumed that the a priori distribution
was uniform according to the indifference principle.
"""
import time
import numpy as np
import matplotlib.pyplot as plt

mu  = 0   # mean
sig = 0.9 #1.3 # standard deviation

def log_likelihood(x, D):
    """
    log likelihood of gaussian

    Parameters
    ----------
    x : 1darray
        array of parameters
    D : int
        dimension of parameter space

    Return
    ------
    likelihood : float
        log likelihood
    """
    
    likelihood = - np.log(2*np.pi*sig**2)*0.5*D - 0.5*np.sum((x - mu)**2)/sig**2  

    return likelihood


def samplig(x, D, bound, step, N_mcmc):
    """
    Sampling a new point of parameter space from
    a uniform distribution as proposal with the
    constraint to go up in likelihood

    Parameters
    ----------
    x : 1darray
        values of parameters and the relative likelihood
        x[:len(x)-1] = parameters
        x[len(x)-1] = likelihood(parameters)
    D : int
        dimension for parameter space, len(x) = D+1
    bound: float
        bounds of parameter space
    step : 1darray
        initial step for each dimension
    N_mcmc : int
        number of montecarlo steps

    Return
    ------
    new_sample : 1darray
        new array replacing x with higher likelihood
    accept : int
        number of moves that have been accepted
    reject : int
        number of moves that have been rejected
    """

    logLmin = x[D] # Worst likelihood
    point = x[:D]  # Point in the parameter space
    accept = 0     # Number of accepted moves
    reject = 0     # Number of rejected moves

    while True:
        # Array initialization
        new_sample = np.zeros(D+1)
        
        # Loop over the components
        for i in range(D):
            #we sample a trial variable with a gaussina distribution
            new_sample[i] = np.random.normal(point[i], step[i])
            
            # If it is out of bound...
            while new_sample[i] <= bound[2*i] or new_sample[i] >= bound[2*i + 1]:
                #...we resample the variable
                new_sample[i] = np.random.normal(point[i], step[i])

        # Computation of the likelihood associated to the new point
        new_sample[D] = log_likelihood(new_sample[:D], D)
        
        reject += 1
        # If the likelihood is greater we accept
        if new_sample[D] > logLmin :
            accept += 1
            reject -= 1 # To avoind one if check
            point[:D] = new_sample[:D]
            

            if accept > N_mcmc: #ACHTUNG
                """
                The samples must be independent. We trust
                that they are after N_mcmc accepted attempts,
                but we should compute the correlation of the D
                chains and the autocorrelation time in order
                to know what to write instead of N_mcmc, which is
                computationally expensive
                """
                break

        # We change the step to go towards a 50% acceptance
        if accept != 0 and reject != 0:
            if accept > reject :
                step *= np.exp(1.0 / accept);
            if accept < reject :
                step /= np.exp(1.0 / reject);

    return new_sample, accept, reject


def nested_samplig(N, D, bound, N_mcmc, tau=1e-6, verbose=False):
    """
    Compute evidence, information and distribution of parameters

    Parameters
    ----------
    N : int
        number of points
    D : int
         dimension for parameter space
    bound: float
        bounds of parameter space
    N_mcmc : int
        number of montecarlo steps
    tau : float
        tollerance, the run stops when the
        variation of evidence is smaller than tau
    verbose : bool, optional
        if True some information are printed during
        the execution to see what is happening

    Retunr
    ------
    calc : dict
        a dictionary which contains several information:
        "evidence"         : logZ,
        "error_lZ"         : error,
        "posterior"        : posterior,
        "worst_likelihood" : np.array(logL_list),
        "prior"            : prior_sample,
        "prior_mass"       : np.array(prior_mass),
        "number_acc"       : accepted,
        "number_rej"       : rejected,
        "number_steps"     : count,
        "log_information"  : np.array(logH_list),
        "list_evidence"    : np.array(logZ_list)

    """

    grid = np.zeros((N, D + 1)) # grid of live points

    prior_mass = [] # Integration variable
    logH_list  = [] # To store the information
    logZ_list  = [] # To store the evidence
    logL_list  = [] # To store the wrost likelihood
    all_value  = [] # To store all sample for the posterior for our parameter

    logH = -np.inf  # ln(Information, initially 0)
    logZ = -np.inf  # ln(Evidence Z, initially 0)

    # Indifference principle, the parameters' priors are uniform
    prior_sample = np.zeros((N, D))
    for i in range(D):
        prior_sample[:, i] = np.random.uniform(bound[2*i], bound[2*i + 1], size=N)

    #initialization of the parameters' values
    grid[:, :D] = prior_sample
    #likelihood initialization
    for i in range(N):
        grid[i, D] = log_likelihood(prior_sample[i,:], D)
    
    # Outermost interval of prior mass
    logwidth = np.log(1.0 - np.exp(-1.0/N))

    count    = 0   # Number of steps
    rejected = 0   # Total rejected steps
    accepted = 0   # Total accepted steps

    while True:
        count += 1                       # We refresh the number of steps
        prior_mass.append(logwidth)      # We store the integration variable

        Lw_idx = np.argmin(grid[:, D])   # Index for the parameters with the worst likelihood, i.e. the smallest one
        logLw = grid[Lw_idx, D]          # Value of the worst likelihood
        
        # Save wrost value for parameters' posterior; np.copy is crucial
        logL_list.append(logLw)
        all_value.append(np.copy(grid[Lw_idx, :D]))

        #np.logaddexp(x, y) = np.log(np.exp(x) + np.exp(y))
        logZnew = np.logaddexp(logZ, logwidth+logLw)

        logZ = logZnew           # We refresh the evidence
        logZ_list.append(logZ)   # We keep the value of the evidence

        # We compute the information and keep it
        logH = np.logaddexp(logH, logwidth + logLw - logZ + np.log(logLw - logZ))
        logH_list.append(logH)

        # New sample used to replace the points we have to delete
        # A good guess for initial step size is the tipical dimension of actual sample
        sampling_step = np.array([np.std(grid[:, jj]) for jj in range(D)])
        new_sample, acc, rej = samplig(grid[Lw_idx], D, bound, sampling_step, N_mcmc)
        accepted += acc # We refresh the total accepted steps
        rejected += rej # We refresh the total rejected steps

        grid[Lw_idx] = new_sample # Replacement
        logwidth -= 1.0/N         # Interval
        
        if verbose :
            # Evidence error computed each time for the print
            error = np.sqrt(np.exp(logH)/N)
            print(f"Iter = {count} acceptance = {accepted/(accepted+rejected):.3f} logZ = {logZ:.3f} error_logZ = {error:.3f} H = {np.exp(logH):.3f} \r", end="")

        if count > 3:
            #break
            if abs((logZ_list[-1] - logZ_list[-2])/logZ_list[-2]) < tau :
                break

    # Evidence error
    error = np.sqrt(np.exp(logH)/N)
    print(f"Iter = {count} acceptance = {accepted/(accepted+rejected):.3f} logZ = {logZ:.3f} error_logZ = {error:.3f} H = {np.exp(logH):.3f} ")
    
    # Compute posterior distributions
    N_resample = int(1e5)
    posterior  = np.zeros((N_resample, D))
    # Compute the weights
    W_sample  = np.exp(np.array(logL_list) + np.array(prior_mass) - logZ_list[-1])
    W_sample /= np.sum(W_sample)
    
    resample_idx = np.random.choice(range(count), N_resample, p=W_sample)
    for i in range(N_resample):
        posterior[i,:] = all_value[resample_idx[i]]
    
    calc = {
            "evidence"         : logZ,
            "error_lZ"         : error,
            "posterior"        : posterior,
            "worst_likelihood" : np.array(logL_list),
            "prior"            : prior_sample,
            "prior_mass"       : np.array(prior_mass),
            "number_acc"       : accepted,
            "number_rej"       : rejected,
            "number_steps"     : count,
            "log_information"  : np.array(logH_list),
            "list_evidence"    : np.array(logZ_list)
    }

    return calc


def plot_hist_par(prior, posterior, D, bound, save=False, show=False):
    """
    Plot of posterior vs prior for all parameters

    Parameters
    ----------
    prior : 2darray
        matrix which contains the prior of all parameters
    posterior : 2darray
        matrix which contains the posterior of all parameters
    D : int
        dimension of the parameter space
    bound : 1darray
        bound for plot
    save : bool, optional
        if True all plots are saved in the current directory
    show : bool, optional
        if True all plots are showed on the screen
    """
    
    for pr, ps, k, b_min, b_max in zip(prior.T, posterior.T, range(D), bound[0::2], bound[1::2]):
        
        fig = plt.figure(k+1)
        plt.title(f"Posterior for {k+1}-th parameter", fontsize=15)
        plt.xlabel("bound", fontsize=15)
        plt.ylabel("Probability density", fontsize=15)
        # normal distribution
        x = np.linspace(b_min, b_max, 1000)
        plt.plot(x, np.exp(-(x - mu)**2/(2*sig**2))/np.sqrt(2*np.pi*sig**2), 'r', label='normal')
        # prior and posterior
        plt.hist(pr, bins=50, density=True, histtype='step', color='blue', label='prior')
        plt.hist(ps, bins=50, density=True, histtype='step', color='black', label="posterior")
        plt.legend(loc='best')
        plt.grid()

        if save :
            plt.savefig(f"parametro{k+1}")
            plt.close(fig)

    if show :
        plt.show()


if __name__ == "__main__":

    np.random.seed(69420)
    
    # Number of points
    N = int(1e3)
    
    # Number of montecarlo steps
    N_mcmc = 20
    
    # Bound for our system: D-dimesional gaussian likelihood
    bound = np.array([-6, 6]*3)
    
    # Dimension of our parameter space
    D = len(bound)//2

    start = time.time()

    NS = nested_samplig(N, D, bound, N_mcmc, verbose=True)

    evidence        = NS["evidence"]
    error_evidence  = NS["error_lZ"]
    posterior_param = NS["posterior"]
    Wrost_L         = NS["worst_likelihood"]
    prior_param     = NS["prior"]
    prior_mass      = NS["prior_mass"]
    acc             = NS["number_acc"]
    rej             = NS["number_rej"]
    number_iter     = NS["number_steps"]
    log_information = NS["log_information"]
    list_evidence   = NS["list_evidence"]

    print(f"Evidence sampling    = {evidence:.3f} +- {error_evidence:.3f}")

    print(f"Number of iterations = {number_iter}")

    acceptance = acc/(acc+rej)
    print(f"Acceptance = {acceptance:.3f}")

    end = time.time() - start

    print(f"Elapsed time = {end//60:.0f} min and {end%60:.0f} s")

    plot_hist_par(prior_param, posterior_param, D, bound, show=True)
    
    plt.figure(1)
    plt.xlabel("Number of iteration", fontsize=15)
    plt.title("Trend of eìevidence and information")
    plt.plot(log_information, 'b', label='Information')
    plt.plot(list_evidence, 'k', label='Evidence')
    plt.legend(loc='best')
    plt.grid()
    plt.show()

File: nested_sampling.jl
#=
Simple code that implements the nested sampling
to compute the evidence of a multidimensional gaussian
=#
using Printf
using Random
using DelimitedFiles

Random.seed!(69420)


function log_likelihood(x, D, vec)
    #=
    log likelihood of gaussian

    Parameters
    ----------
    x : array or matrix
        array of parameters
    D : int
        dimension of parameter space
    vec : bool
        if True x must be a matrix NxD, this is convenient to
        initialize the likelihood at first step of nested_sampling
        if False x must be a D-dimensional array

    Return
    ------
    likelihood : array or float
        log likelihood of gaussian
        list if vec = True
        float if vec = False
    =#
    if vec == true
        N = length(x[:, 1])
        log_likelihood = Float64[]
        for i in 1:N
            v = x[i, :]
            g = - 0.5*D*log(2*pi) - 0.5*sum(v.^2)
            push!(log_likelihood, g)
        end
    else
        log_likelihood = - 0.5*D*log(2*pi) - 0.5*sum(x.^2)
    end
    
    return log_likelihood
    
end


function sampling(x, D, bound, thr)
    #=
    Sampling a new point of parameter space from
    a uniform distribution as proposal with the
    constraint to go up in likelihood

    Parameters
    ----------
    x : 1darray
        values of parameters and the relative likelihood
        x[1:D] = parameters
        x[D-1] = likelihood(parameters)
    D : int
        dimension for parameter space, len(x) = D+1
    bound: float
        bounds of parameter space
    thr : int
        threshold for acceptance for indipendent sampling

    Return
    ------
    new_sample : 1darray
        new array replacing x with higher likelihood
    accept : int
        number of moves that have been accepted
    reject : int
        number of moves that have been rejected
    =#
    local new_sample, accept, reject

    logLmin = x[D+1] #worst likelihood
    point = x[1:D]   #point in the parameter space
    step = 0.1       #initial step of the algorithm, to set
    accept = 0       #number of accepted moves
    reject = 0       #number of rejected moves
    
    while true
        #array initialization
        new_sample = Float64[]
        for i in 1:D+1
            push!(new_sample, 0.0)
        end
        #loop over the components
        for i in 1:D
            #we sample a trial variable
            draw =  step * (2*rand(Float64) - 1)
            new_sample[i] = point[i] + draw
            #...we resample the variable
            while abs(new_sample[i]) > bound
                #...we resample the variable
                draw =  step * (2*rand(Float64) - 1)
                new_sample[i] = point[i] + draw
            end
        end
        #computation of the likelihood associated to the new point
        new_sample[D+1] = log_likelihood(new_sample[1:D], D, false)

        #if the likelihood is smaller than before we reject
        if new_sample[D+1] < logLmin
            reject += 1
        
        #if greater we accept
        elseif new_sample[D+1] > logLmin
            accept += 1
            point[1:D] = new_sample[1:D]

            if accept > thr #ACHTUNG
                #=
                the samples must be independent. We trust
                that they are after thr accepted attempts,
                but we should compute the correlation of the D
                chains and the autocorrelation time in order
                to know what to write instead of thr, which is
                computationally expensive
                =#
                break
            end
        end
        
        # We change the step to go towards a 50% acceptance
        if accept != 0 && reject != 0
            if accept > reject 
                step *= exp(1.0 / accept)
            elseif accept < reject 
                step /= exp(1.0 / reject)
            end
        end

    end
    
    return new_sample, accept, reject

end


function log_sum(x, y)
    #=
    compute log(exp(x) + exp(y))
    numerically stable
    
    Parameter
    ---------
    x : float
        argument of first exp
    y : float
        argument of second exp
    
    Return
    ------
    sum : float
        log(exp(x) + exp(y))
    =#
    if x > y
        sum = x + log(1 + exp(y - x)) 
    else
        sum = y + log(1 + exp(x - y))
    end
    
    return sum
end


function nested_samplig(N, D, bound, tau, thr, verbose=false)
    #=
    """
    Compute evidence, information and distribution of parameters

    Parameters
    ----------
    N : int
        number of points
    D : int
         dimension for parameter space
    bound: float
        bounds of parameter space
    tau : float
        tollerance, the run stops when the
        variation of evidence is smaller than tau
    thr : int
        threshold for acceptance for indipendent sampling
    verbose : bool, optional
        if True some information are printed during
        the execution to see what is happening

    Retunr
    ------
    calc : dict
        a dictionary which contains several information:
        "evidence"        => logZ,
        "error_lZ"        => error,
        "posterior"       => grid[:, :D],
        "likelihood"      => grid[:,  D],
        "prior"           => prior_sample,
        "prior_mass"      => np.array(prior_mass),
        "number_acc"      => accepted,
        "number_rej"      => rejected,
        "number_steps"    => Iter,
        "log_information" => np.array(logH_list),
        "list_evidence"   => np.array(logZ_list)
    =#
    local logZ, error
    grid = rand(Float64, (N, D+1))
    
    prior_mass = Float64[] #integration variable
    logH_list  = Float64[] #we keep the information
    logZ_list  = Float64[] #we keep the evidence

    logH = -Inf  # ln(Information, initially 0)
    logZ = -Inf  # ln(Evidence Z, initially 0)

    #indifference principle, the parameters' priors are uniform
    prior_sample = bound .* (2 .* rand(Float64, (N, D)) .- 1)

    #initialization of the parameters' values
    grid[:, 1:D] = prior_sample
    #likelihood initialization
    grid[:, D+1] = log_likelihood(prior_sample, D, true)
    
    # Outermost interval of prior mass
    logwidth = log(1.0 - exp(-1.0/N))

    Iter     = 0   #number of steps
    rejected = 0   #total rejected steps
    accepted = 0   #total accepted steps
    
    while true
        
        Iter += 1                        #we refresh the number of steps
        push!(prior_mass, logwidth)      #we keep the integration variable

        Lw_idx = argmin(grid[:, D+1])    #Index for the parameters with the
                                         #worst likelihood, i.e. the smallest one
        logLw = grid[Lw_idx, D+1]        #Value of the worst likelihood
        
        logZnew = log_sum(logZ, logwidth+logLw)

        logZ = logZnew           #we refresh the evidence
        push!(logZ_list, logZ)   #we keep the value of the evidence

        #we compute the information and keep it
        logH = log_sum(logH, logwidth + logLw - logZ + log(logLw - logZ))
        push!(logH_list, logH)

        #new sample used to replace the points we have to delete
        new_sample, acc, rej = sampling(grid[Lw_idx, :], D, bound, thr)
        accepted += acc #we refresh the total accepted steps
        rejected += rej #we refresh the total rejected steps
        
        grid[Lw_idx, :] = new_sample #replacement
        logwidth -= 1.0/N            #interval shrinking
        
        if verbose
            error = sqrt(exp(logH)/N)
            Acc = accepted/(accepted+rejected)
            @printf("Iter = %d, logZ = %0.5f +- %0.5f, acc = %0.3f \n", Iter, logZ, error, Acc)
        end
        
        if Iter > 3
            if abs(logZ_list[end] - logZ_list[end-1]) < tau
                break
            end
        end
    end
    #evidence error
    error = sqrt(exp(logH)/N)
    
    calc = Dict(
                "evidence"        => logZ,
                "error_lZ"        => error,
                "posterior"       => grid[:, 1:D],
                "likelihood"      => grid[:, D+1],
                "prior"           => prior_sample,
                "prior_mass"      => prior_mass,
                "number_acc"      => accepted,
                "number_rej"      => rejected,
                "number_steps"    => Iter,
                "log_information" => logH_list,
                "list_evidence"   => logZ_list
               )
    
    return calc
    
end

#number of point
N = Int(1e4)
#dimesion of space
D = 50
#integration limit, in these units it is the number of standard deviations
bound = 6
#tollerance of algorithm
tau = 1e-6
#threshold for acceptance must be tuned (17 D=50), (40 D=3)
thr = 17

@time NS = nested_samplig(N, D, bound, tau, thr, true)

evidence        = NS["evidence"]
error_evidence  = NS["error_lZ"]
posterior_param = NS["posterior"]
likelihood      = NS["likelihood"]
prior_param     = NS["prior"]
prior_mass      = NS["prior_mass"]
acc             = NS["number_acc"]
rej             = NS["number_rej"]
number_iter     = NS["number_steps"]
log_information = NS["log_information"]
list_evidence   = NS["list_evidence"]

#analitc result
logZ_v = -D*log(2*bound)

@printf("Evidence sampling    = %0.5f +- %0.5f \n", evidence, error_evidence)
@printf("Theoretical evidence = %0.5f \n", logZ_v)

@printf("Number of iterations = %d \n", number_iter)

acceptance = acc/(acc+rej)
@printf("Acceptance = %0.3f \n", acceptance)

#save some information
writedlm("param.txt", [N, D])
writedlm("prior.txt", prior_param)
writedlm("poste.txt", posterior_param)

    

